File: EarthquakeList/src/main/java/com/jseo/earthquakelist/actors/DataParser.java
package com.jseo.earthquakelist.actors;

/**
 * parses and returns usable data
 */
public abstract class DataParser {

    private OnParseCompleteListener mOnCompleteListener;
    private Object mData;

    interface OnParseCompleteListener {
        void onParseComplete(boolean isSuccess, Object parsedData);
    }

    DataParser() {
        mOnCompleteListener = null;
        mData = null;
    }

    DataParser(Object data) {
        mOnCompleteListener = null;
        mData = data;
    }

    public void setOnParseCompleteListener(OnParseCompleteListener listener) {
        mOnCompleteListener = listener;
    }

    public OnParseCompleteListener getOnParseCompleteListener() {
        return mOnCompleteListener;
    }

    /**
     * which input reader?
     * @param data probably an input reader
     */
    public void setData(Object data) {
        mData = data;
    }

    public Object getData() { return mData; }

    //let derived class implement how to parse data
    abstract void parse();


}

File: EarthquakeList/src/main/java/com/jsbomb/earthquakelist/actors/DataParser.kt
package com.jsbomb.earthquakelist.actors

/**
 * parses and returns usable data
 */
abstract class DataParser(
    private var data: Any? = null,
    private var onCompleteListener: OnParseCompleteListener? = null
) {

    interface OnParseCompleteListener {
        fun onParseComplete(isSuccess: Boolean, parsedData: Any?)
    }

    fun setOnParseCompleteListener(listener: OnParseCompleteListener) {
        onCompleteListener = listener
    }

    fun getOnParseCompleteListener(): OnParseCompleteListener? = onCompleteListener

    fun setData(data: Any) {
        this.data = data
    }

    fun getData(): Any? = data

    abstract fun parse()
}

